source/extensions/transport_sockets/tls/context_impl.h

```diff
@@ -264,9 +256,11 @@ class ServerContextImpl : public ContextImpl, public Envoy::Ssl::ServerContext {
                          unsigned int inlen);
   int sessionTicketProcess(SSL* ssl, uint8_t* key_name, uint8_t* iv, EVP_CIPHER_CTX* ctx,
                            HMAC_CTX* hmac_ctx, int encrypt);
-  OcspStapleAction ocspStapleAction(const ServerContextImpl::TlsContext& ctx,
+  bool isClientOcspCapable(SSL* ssl);
+  const ContextImpl::CertContext& certificateContext(X509* cert);
```

Comment:
These are specific to our implementation right? Should we consider documenting any new method(s)? This would then also be a good  place to call out the assumptions around inputs (eg. when this asserts). Tangentially related, I'm curious, we `RELEASE_ASSERT` for certain inputs of `cert` here in the implementation of this, would this be a candidate for returning an `absl::status_or` instead and handling it in `ServerContextImpl::handleOcspStapling`? If this is truly invariant, then probably not, but in that case I'm wondering if there are tests in place around verifying that.

Reply:
> These are specific to our implementation right?

They are. Although upstream has `isClientOcspCapable(), here: https://github.com/envoyproxy/envoy/blob/release/v1.16/source/extensions/transport_sockets/tls/context_impl.cc#L1391` which is similar in intent, but relies on BoringSSL api. In this case it's an API that doesn't have an OpenSSL analogue (parsing of SSL handshake), so I didn't bother trying to extend bssl wrapper with an implementation for it.

> Should we consider documenting any new method(s)?

These two are trivial, and I tried to use method names that are self-explanatory. I added comments to new/modified data-structures (`certificateContext()` retrieves a reference to a Maistra-specific one, for example) however.

> Tangentially related, I'm curious, we RELEASE_ASSERT for certain inputs of cert here

Not entirely sure if you are talking about this: https://github.com/maistra/envoy/pull/86/files#diff-2b39f212c2f760ae547351a46f92ad08c0ec84c556d678ebe9388b9d008e7313R1290. If so: the assert here is b/c we should never ever *not* match a certificate: it's selected by OpenSSL from certs configured for a given TLS context (here's the lookup map being populated: https://github.com/maistra/envoy/pull/86/files#diff-2b39f212c2f760ae547351a46f92ad08c0ec84c556d678ebe9388b9d008e7313R313, a bit earlier there the cert chain itself is being loaded into the SSL_CTX). Not finding a match wouldn't be a runtime error, but a symptom of something being seriously wrong with the implementation.

> If this is truly invariant, then probably not, but in that case I'm wondering if there are tests in place around verifying that.

No additional tests, but I don't think we generally add tests to verify asserts (again, this isn't to catch a runtime issue).

Reply:
Thank you for explaining, that is very helpful. Based on that I can see `RELEASE_ASSERT` is the right choice here, and covering it in a test doesn't make sense (as it is invariant). 
As for doc comments, personally I tend try to err on the "generous" side (assuming naivity of future readers) but that is subjective, so up to you.